<div style="background: black; padding: 10px 250px"><img src="https://www.veldikompetens.se/wp-content/themes/consid/static/icons/VeldiKompetens_Logo_Web_Negative.svg" title="Veldi kompetens" /></div>

<hr><h1><center>Exercise 3a - Pandas Basics </center></h1>

<h3>Instructions </h3>

<p>In this exercise you will learn about the basics of Pandas; Loading a dataset, rows, columns, showing header, printing a set of lines and for example printing a row or a column. The dataset we will be using in this exercise and forwards is about different types of chocolates and how professionals have rated them. Lets get to know the dataset! 

If you want to read some more on the dataset you can find information here https://www.kaggle.com/rtatman/chocolate-bar-ratings</p>

<h3> 1. Setup </h3>

Neccesary libraries:

In [1]:
import numpy as np
import pandas as pd

<h3> 2. Pandas Basics</h3>
<h4> 2.1 Loading the dataset </h4>
<p> Note: it is a common practice to name the variable df which stands for Data Frame. Since you will be using this variable A LOT it is useful to have a concise name just as we do with numpy and pandas. Lets take a look!  </p>
<p>IMPORTANT: You need to download the dataset first and put it into the same folder as the current notebook you are working in to run the code below. Dataset can be found on itslearning </p>

In [5]:
df = pd.read_csv("flavors_of_cacao.csv", na_values = [" "])
df


Company \n(Maker-if known) Specific Bean Origin\nor Bar Name   REF  \
0                      A. Morin                       Agua Grande  1876   
1                      A. Morin                             Kpime  1676   
2                      A. Morin                            Atsane  1676   
3                      A. Morin                             Akata  1680   
4                      A. Morin                            Quilla  1704   
...                         ...                               ...   ...   
1790                     Zotter                              Peru   647   
1791                     Zotter                             Congo   749   
1792                     Zotter                      Kerala State   749   
1793                     Zotter                      Kerala State   781   
1794                     Zotter                Brazil, Mitzi Blue   486   

      Review\nDate Cocoa\nPercent Company\nLocation  Rating Bean\nType  \
0             2016            63%            France    3.75              
1             2015            70%            France    2.75              
2             2015            70%            France    3.00              
3             2015            70%            France    3.50              
4             2015            70%            France    3.50              
...            ...            ...               ...     ...        ...   
1790          2011            70%           Austria    3.75              
1791          2011            65%           Austria    3.00  Forastero   
1792          2011            65%           Austria    3.50  Forastero   
1793          2011            62%           Austria    3.25              
1794          2010            65%           Austria    3.00              

     Broad Bean\nOrigin  
0              Sao Tome  
1                  Togo  
2                  Togo  
3                  Togo  
4                  Peru  
...                 ...  
1790               Peru  
1791              Congo  
1792              India  
1793              India  
1794             Brazil  

[1795 rows x 9 columns]

Looks nice right? Pandas visualizes the csv file and gives  you a ton of information just from reading the csv file. Truly an amazing library.

<h4>2.2 Some useful commands </h4>

In [6]:
print(f"The type is: {type(df)} \n")
print(f"The datatypes for each column: \n {df.dtypes}")

The type is: <class 'pandas.core.frame.DataFrame'> 

The datatypes for each column: 
 Company \n(Maker-if known)            object
Specific Bean Origin\nor Bar Name     object
REF                                    int64
Review\nDate                           int64
Cocoa\nPercent                        object
Company\nLocation                     object
Rating                               float64
Bean\nType                            object
Broad Bean\nOrigin                    object
dtype: object


<p>Before we proceed ensure that you have examined the dataset above and have a feel for what type of data value should be present in each column. Now lets move onto learning some useful commands! </p>

In [16]:
# Note; Might look really ugly but the notebook cant put everything on one line! The \ is a separator to signify same print
# But new line
#print("From the bottom: ",df.tail(1), "\n")
#print("From the top: ", df.head(1), "\n")

# Can also do as below to make use of Pandas awesome integration with jupyter and get a real table!
df.head(100)
#df.tail(0)

Company \n(Maker-if known) Specific Bean Origin\nor Bar Name   REF  \
0                    A. Morin                       Agua Grande  1876   
1                    A. Morin                             Kpime  1676   
2                    A. Morin                            Atsane  1676   
3                    A. Morin                             Akata  1680   
4                    A. Morin                            Quilla  1704   
..                        ...                               ...   ...   
95                        Ara                        Madagascar  1375   
96                        Ara                           Chiapas  1379   
97                        Ara                          Equateur  1379   
98                        Ara                        Trincheras  1379   
99                      Arete                          San Juan  1724   

    Review\nDate Cocoa\nPercent Company\nLocation  Rating  Bean\nType  \
0           2016            63%            France    3.75               
1           2015            70%            France    2.75               
2           2015            70%            France    3.00               
3           2015            70%            France    3.50               
4           2015            70%            France    3.50               
..           ...            ...               ...     ...         ...   
95          2014            75%            France    3.00  Trinitario   
96          2014            72%            France    2.50               
97          2014            75%            France    2.75               
98          2014            75%            France    3.00               
99          2016            70%            U.S.A.    3.50  Trinitario   

   Broad Bean\nOrigin  
0            Sao Tome  
1                Togo  
2                Togo  
3                Togo  
4                Peru  
..                ...  
95         Madagascar  
96             Mexico  
97            Ecuador  
98          Venezuela  
99           Trinidad  

[100 rows x 9 columns]

Something that wont be showcased here but is worth pointing out is that pandas also has a functionality called df.to_numpy() which makes ur entire dataframe into a numpy array. Its quick and cheap in terms of computational cost for inputs of the same type and is ideally used for usage on say all columns sharing type. Anyways, onwards we go!

We just saw the head (i.e the top) and the tail (i.e the bottom) of the dataframe. To access all values (indexes or rows) and columns there are also commands for that purpose as can be seen below

In [19]:
#print(df.columns)
print(df.index)

RangeIndex(start=0, stop=1795, step=1)


Next up is the most descriptive functionality of all: describe

In [20]:
# Does not include every column as only descriptive summaries of descriptive values can be showcased
print(df.describe())

               REF  Review\nDate       Rating
count  1795.000000   1795.000000  1795.000000
mean   1035.904735   2012.325348     3.185933
std     552.886365      2.927210     0.478062
min       5.000000   2006.000000     1.000000
25%     576.000000   2010.000000     2.875000
50%    1069.000000   2013.000000     3.250000
75%    1502.000000   2015.000000     3.500000
max    1952.000000   2017.000000     5.000000


There are several ways of sorting your data, either by index or by column values. One of the more common ones is df.sort_values(). Note we use copy here just as we do with numpy!

In [34]:
sorted_df = df.copy().sort_values(("REF"))
sorted_df

Company \n(Maker-if known) Specific Bean Origin\nor Bar Name   REF  \
1326                     Pralus                          Trinidad     5   
1327                     Pralus         Madagascar, Nosy Be Isle.     5   
878               Jacque Torres               Trinatario Treasure     5   
1325                     Pralus               Sao Tome & Principe     5   
1177        Neuhaus (Callebaut)                          Sao Tome    15   
...                         ...                               ...   ...   
1760                    Xocolla          Hispaniola, batch 170104  1948   
1759                    Xocolla           Sambirano, batch 170102  1948   
958                 Letterpress          La Masica, Batch 7, FHIA  1952   
535                 Dick Taylor                    Lanquin Estate  1952   
959                 Letterpress      Maya Mtn, Batch 18, Heirloom  1952   

      Review\nDate Cocoa\nPercent Company\nLocation  Rating  Bean\nType  \
1326          2006            75%            France    3.50  Trinitario   
1327          2006            75%            France    3.75     Criollo   
878           2006            71%            U.S.A.    2.00  Trinitario   
1325          2006            75%            France    3.50   Forastero   
1177          2006            75%           Belgium    2.75   Forastero   
...            ...            ...               ...     ...         ...   
1760          2017            70%            U.S.A.    2.50               
1759          2017            70%            U.S.A.    2.75               
958           2017            70%            U.S.A.    3.50               
535           2017            70%            U.S.A.    3.75               
959           2017            70%            U.S.A.    3.25               

       Broad Bean\nOrigin  
1326             Trinidad  
1327           Madagascar  
878                 Ghana  
1325  Sao Tome & Principe  
1177             Sao Tome  
...                   ...  
1760   Dominican Republic  
1759           Madagascar  
958              Honduras  
535             Guatemala  
959                Belize  

[1795 rows x 9 columns]

<h4>2.3 Getting values </h4>

In [22]:
# Returns the values from specified column
df["REF"]

0       1876
1       1676
2       1676
3       1680
4       1704
        ... 
1790     647
1791     749
1792     749
1793     781
1794     486
Name: REF, Length: 1795, dtype: int64

In [29]:
# Each row from 0 to 1
df[0:4]

Company \n(Maker-if known) Specific Bean Origin\nor Bar Name   REF  \
0                   A. Morin                       Agua Grande  1876   
1                   A. Morin                             Kpime  1676   
2                   A. Morin                            Atsane  1676   
3                   A. Morin                             Akata  1680   

   Review\nDate Cocoa\nPercent Company\nLocation  Rating Bean\nType  \
0          2016            63%            France    3.75              
1          2015            70%            France    2.75              
2          2015            70%            France    3.00              
3          2015            70%            France    3.50              

  Broad Bean\nOrigin  
0           Sao Tome  
1               Togo  
2               Togo  
3               Togo

Two other common ways of accessing data is by usage of loc and iloc. In pandas loc is used to get rows and/or columns with particular label, whilst iloc gets them via integer locations or index

In [25]:
# Get a specific row by the INDEX it is at (Note when we loaded the data pandas generated indexes for each instance) 
sorted_df.iloc[3]

Company \n(Maker-if known)                        Pralus
Specific Bean Origin\nor Bar Name    Sao Tome & Principe
REF                                                    5
Review\nDate                                        2006
Cocoa\nPercent                                       75%
Company\nLocation                                 France
Rating                                               3.5
Bean\nType                                     Forastero
Broad Bean\nOrigin                   Sao Tome & Principe
Name: 1325, dtype: object

In [30]:
# Get a specific row from the label (this means that originally this object was given index 3 as can be seen by name)
sorted_df.loc[3]

Company \n(Maker-if known)           A. Morin
Specific Bean Origin\nor Bar Name       Akata
REF                                      1680
Review\nDate                             2015
Cocoa\nPercent                            70%
Company\nLocation                      France
Rating                                    3.5
Bean\nType                                   
Broad Bean\nOrigin                       Togo
Name: 3, dtype: object

<h4>2.4 Time for some Q/A! </h4> 

In [32]:
#Q1: Output the first three rows in sorted_df and df respectively
# here output first three of df
df[0:3]


Company \n(Maker-if known) Specific Bean Origin\nor Bar Name   REF  \
0                   A. Morin                       Agua Grande  1876   
1                   A. Morin                             Kpime  1676   
2                   A. Morin                            Atsane  1676   

   Review\nDate Cocoa\nPercent Company\nLocation  Rating Bean\nType  \
0          2016            63%            France    3.75              
1          2015            70%            France    2.75              
2          2015            70%            France    3.00              

  Broad Bean\nOrigin  
0           Sao Tome  
1               Togo  
2               Togo

Expected output: 
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Company&nbsp; (Maker-if known)</th>
      <th>Specific Bean Origin or Bar Name</th>
      <th>REF</th>
      <th>Review Date</th>
      <th>Cocoa Percent</th>
      <th>Company Location</th>
      <th>Rating</th>
      <th>Bean Type</th>
      <th>Broad Bean Origin</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>A. Morin</td>
      <td>Agua Grande</td>
      <td>1876</td>
      <td>2016</td>
      <td>63%</td>
      <td>France</td>
      <td>3.75</td>
      <td></td>
      <td>Sao Tome</td>
    </tr>
    <tr>
      <th>1</th>
      <td>A. Morin</td>
      <td>Kpime</td>
      <td>1676</td>
      <td>2015</td>
      <td>70%</td>
      <td>France</td>
      <td>2.75</td>
      <td></td>
      <td>Togo</td>
    </tr>
    <tr>
      <th>2</th>
      <td>A. Morin</td>
      <td>Atsane</td>
      <td>1676</td>
      <td>2015</td>
      <td>70%</td>
      <td>France</td>
      <td>3.00</td>
      <td></td>
      <td>Togo</td>
    </tr>
  </tbody>
</table>

In [39]:
# here output first three of sorted_df
sorted_df.iloc[0:3]


Company \n(Maker-if known) Specific Bean Origin\nor Bar Name  REF  \
1326                     Pralus                          Trinidad    5   
1327                     Pralus         Madagascar, Nosy Be Isle.    5   
878               Jacque Torres               Trinatario Treasure    5   

      Review\nDate Cocoa\nPercent Company\nLocation  Rating  Bean\nType  \
1326          2006            75%            France    3.50  Trinitario   
1327          2006            75%            France    3.75     Criollo   
878           2006            71%            U.S.A.    2.00  Trinitario   

     Broad Bean\nOrigin  
1326           Trinidad  
1327         Madagascar  
878               Ghana

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Company(Maker-if known)</th>
      <th>Specific Bean Origin or Bar Name</th>
      <th>REF</th>
      <th>Review Date</th>
      <th>Cocoa Percent</th>
      <th>Company Location</th>
      <th>Rating</th>
      <th>Bean Type</th>
      <th>Broad Bean Origin</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>882</th>
      <td>Jacque Torres</td>
      <td>Trinatario Treasure</td>
      <td>5.0</td>
      <td>2006.0</td>
      <td>71%</td>
      <td>U.S.A.</td>
      <td>2.00</td>
      <td>Trinitario</td>
      <td>Ghana</td>
    </tr>
    <tr>
      <th>1330</th>
      <td>Pralus</td>
      <td>Trinidad</td>
      <td>5.0</td>
      <td>2006.0</td>
      <td>75%</td>
      <td>France</td>
      <td>3.50</td>
      <td>Trinitario</td>
      <td>Trinidad</td>
    </tr>
    <tr>
      <th>1331</th>
      <td>Pralus</td>
      <td>Madagascar, Nosy Be Isle.</td>
      <td>5.0</td>
      <td>2006.0</td>
      <td>75%</td>
      <td>France</td>
      <td>3.75</td>
      <td>Criollo</td>
      <td>Madagascar</td>
    </tr>
  </tbody>
</table>

In [41]:
#Q2: Get the Broad Bean origin for the third item from the top from the sorted_df (i.e Ghana)
# Hint: Use iloc to get a specific row from the top, syntax should be sorted_df.iloc[smthing][column name]
sorted_df.iloc[2]["Broad Bean\nOrigin"]

'Ghana'

Great job! We are now done with the basic introduction to Pandas! Lets proceed to the next exercise where we will learn to work some more with the dataset and reason a bit more on it